lib/core/test/unit/Cardano/Pool/DB/Properties.hs

```diff
@@ -0,0 +1,296 @@
+    prop db' = liftIO $ do
+        forM_ pairs $ \(pool, slot) -> do
+            let err = ErrSlotAlreadyExists slot
+            runExceptT (putPoolProduction db' slot pool) `shouldReturn` Right ()
+            runExceptT (putPoolProduction db' slot pool) `shouldReturn` Left err
+            runExceptT (putPoolProduction db slot pool) `shouldReturn` Right ()
+            runExceptT (putPoolProduction db slot pool) `shouldReturn` Left err
+
+-- | Rolling back wipes out pool production statistics after the rollback point.
+prop_rollbackPools
+    :: DBLayer IO
+    -> StakePoolsFixture
+    -> SlotId
+    -> Property
+prop_rollbackPools db f@(StakePoolsFixture pairs _) sl =
+    monadicIO prop
+  where
+    prop = do
+        (beforeRollback, afterRollback) <- liftIO $ do
+            forM_ pairs $ \(pool, slot) ->
+                runExceptT $ putPoolProduction db slot pool
+            before <- map fst <$> allPoolProduction db f
+            rollbackTo db sl
+            after <- map fst <$> allPoolProduction db f
+            pure (before, after)
+
+        monitor $ counterexample $ unlines
+            [ "Rollback point:    " <> showSlot sl
+            , "Production before: " <> unwords (map showSlot beforeRollback)
+            , "Production after:  " <> unwords (map showSlot afterRollback)
+            ]
+        monitor $ classify (any (> sl) beforeRollback) "something to roll back"
+        monitor $ classify (all (<= sl) beforeRollback) "nothing to roll back"
+
+        assert $ all (<= sl) afterRollback
+
+    showSlot (SlotId epoch slot) = show epoch ++ "." ++ show slot
+
+-- | Can read pool production only for a given epoch
+prop_readPoolNoEpochLeaks
+    :: DBLayer IO
+    -> StakePoolsFixture
+    -> Property
+prop_readPoolNoEpochLeaks db (StakePoolsFixture pairs _) =
+    monadicIO (setup >> prop)
+  where
+    slotPartition = L.groupBy ((==) `on` epochNumber)
+        $ L.sortOn epochNumber
+        $ map snd pairs
+    epochGroups = L.zip (uniqueEpochs pairs) slotPartition
+    setup = liftIO $ cleanDB db
+    prop = liftIO $ do
+        forM_ pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db slot pool
+        forM_ epochGroups $ \(epoch, slots) -> do
+            slots' <-
+                (Set.fromList . concat . Map.elems) <$> readPoolProduction db epoch
+            slots' `shouldBe` (Set.fromList slots)
+
+-- | Read pool production satisfies conditions after consecutive
+-- 1-slot-depth rollbacks
+prop_readPoolCondAfterDeterministicRollbacks
+    :: (Map PoolId [SlotId] -> Expectation)
+    -> DBLayer IO
+    -> StakePoolsFixture
+    -> Property
+prop_readPoolCondAfterDeterministicRollbacks cond db (StakePoolsFixture pairs _) =
+    monadicIO (setup >> prop)
+  where
+    setup = liftIO $ cleanDB db
+    slots = map snd pairs
+    prop = liftIO $ do
+        forM_ pairs $ \(pool, slot) ->
+            unsafeRunExceptT $ putPoolProduction db slot pool
+        forM_ slots $ \slot -> do
```

Comment:
This should probably be `reverse slots` :thinking: ?

Reply:
I checked it. `slots` (without reverse) goes through every slot from the latest one down to oldest one and apply rollback as I intended